File: android/src/main/java/com/capacitor/mediaviewer/HlsPlaylistParser.java
package com.capacitor.mediaviewer;

import android.util.Log;
import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class HlsPlaylistParser {

    private static final String TAG = "HlsPlaylistParser";

    public static List<QualityVariant> parseMasterPlaylist(String playlistUrl) {
        List<QualityVariant> variants = new ArrayList<>();

        try {
            URL url = new URL(playlistUrl);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("GET");
            connection.setConnectTimeout(5000);
            connection.setReadTimeout(5000);

            InputStream inputStream = connection.getInputStream();
            BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream));

            String line;
            String currentLabel = null;
            String currentVariantUrl = null;
            int currentWidth = 0;
            int currentHeight = 0;

            Pattern resolutionPattern = Pattern.compile("RESOLUTION=(\\d+)x(\\d+)");
            Pattern bandwidthPattern = Pattern.compile("BANDWIDTH=(\\d+)");

            while ((line = reader.readLine()) != null) {
                line = line.trim();

                if (line.startsWith("#EXT-X-STREAM-INF:")) {
                    currentWidth = 0;
                    currentHeight = 0;
                    currentLabel = null;

                    Matcher resolutionMatcher = resolutionPattern.matcher(line);
                    if (resolutionMatcher.find()) {
                        currentWidth = Integer.parseInt(resolutionMatcher.group(1));
                        currentHeight = Integer.parseInt(resolutionMatcher.group(2));
                        currentLabel = currentHeight + "p"; // e.g., 1080p
                    }

                    if (currentLabel == null) {
                        Matcher bandwidthMatcher = bandwidthPattern.matcher(line);
                        if (bandwidthMatcher.find()) {
                            int bandwidth = Integer.parseInt(bandwidthMatcher.group(1));
                            if (bandwidth < 500000) {
                                currentLabel = "SD";
                            } else if (bandwidth < 2000000) {
                                currentLabel = "HD";
                            } else {
                                currentLabel = "Full HD";
                            }
                        }
                    }
                } else if (!line.startsWith("#") && !line.isEmpty()) {
                    currentVariantUrl = resolveUrl(playlistUrl, line);

                    if (currentVariantUrl != null && currentLabel != null) {
                        QualityVariant variant = new QualityVariant();
                        variant.label = currentLabel;
                        variant.url = currentVariantUrl;
                        variant.width = currentWidth;
                        variant.height = currentHeight;

                        variants.add(variant);

                        // Reset flags
                        currentLabel = null;
                        currentVariantUrl = null;
                    }
                }
            }

            reader.close();
            inputStream.close();
            connection.disconnect();
        } catch (Exception e) {
            Log.e(TAG, "Error parsing HLS playlist: " + e.getMessage());
        }

        return variants;
    }

    private static String resolveUrl(String baseUrl, String relativeUrl) {
        try {
            if (relativeUrl.startsWith("http://") || relativeUrl.startsWith("https://")) {
                return relativeUrl;
            }

            URL base = new URL(baseUrl);
            URL resolved = new URL(base, relativeUrl);
            return resolved.toString();
        } catch (Exception e) {
            return relativeUrl;
        }
    }

    public static boolean isHlsUrl(String url) {
        return url != null && (url.endsWith(".m3u8") || url.contains(".m3u8"));
    }
}

File: ios/Plugin/HlsPlaylistParser.swift
import Foundation
import AVFoundation

class HlsPlaylistParser {
    static func parseMasterPlaylist(_ playlistUrl: String, completion: @escaping ([QualityVariant]) -> Void) {
        guard let url = URL(string: playlistUrl) else {
            completion([])
            return
        }
        
        // Use background queue for network request
        DispatchQueue.global(qos: .userInitiated).async {
            var variants: [QualityVariant] = []
            
            do {
                // Fetch the playlist content
                let playlistContent = try String(contentsOf: url, encoding: .utf8)
                let lines = playlistContent.components(separatedBy: .newlines)
                
                var currentResolution: String?
                var currentBandwidth: String?
                var currentVariantUrl: String?
                var currentWidth = 0
                var currentHeight = 0
                
                // Patterns for parsing HLS playlist
                let resolutionPattern = try NSRegularExpression(pattern: "RESOLUTION=(\\d+)x(\\d+)", options: [])
                let bandwidthPattern = try NSRegularExpression(pattern: "BANDWIDTH=(\\d+)", options: [])
                
                for line in lines {
                    let trimmedLine = line.trimmingCharacters(in: .whitespaces)
                    
                    if trimmedLine.hasPrefix("#EXT-X-STREAM-INF:") {
                        currentWidth = 0
                        currentHeight = 0
                        currentResolution = nil
                        
                        // Extract resolution
                        let resolutionRange = NSRange(location: 0, length: trimmedLine.utf16.count)
                        if let resolutionMatch = resolutionPattern.firstMatch(in: trimmedLine, options: [], range: resolutionRange) {
                            if let widthRange = Range(resolutionMatch.range(at: 1), in: trimmedLine),
                               let width = Int(trimmedLine[widthRange]),
                               let heightRange = Range(resolutionMatch.range(at: 2), in: trimmedLine),
                               let height = Int(trimmedLine[heightRange]) {
                                currentWidth = width
                                currentHeight = height
                                currentResolution = "\(height)p"
                            }
                        }
                        
                        // Extract bandwidth if no resolution
                        if currentResolution == nil {
                            if let bandwidthMatch = bandwidthPattern.firstMatch(in: trimmedLine, options: [], range: resolutionRange),
                               let bandwidthRange = Range(bandwidthMatch.range(at: 1), in: trimmedLine),
                               let bandwidth = Int(trimmedLine[bandwidthRange]) {
                                if bandwidth < 500000 {
                                    currentResolution = "SD"
                                } else if bandwidth < 2000000 {
                                    currentResolution = "HD"
                                } else {
                                    currentResolution = "Full HD"
                                }
                                currentBandwidth = String(bandwidth)
                            }
                        }
                        
                        // If still no resolution, use default
                        if currentResolution == nil {
                            currentResolution = "Auto"
                        }
                    } else if !trimmedLine.hasPrefix("#") && !trimmedLine.isEmpty {
                        // This is a URL line
                        currentVariantUrl = resolveUrl(baseUrl: url, relativeUrl: trimmedLine)
                        
                        if let variantUrl = currentVariantUrl, let resolution = currentResolution {
                            let variant = QualityVariant()
                            variant.label = resolution
                            variant.url = variantUrl
                            variant.width = currentWidth
                            variant.height = currentHeight
                            variants.append(variant)
                            
                            // Reset for next variant
                            currentResolution = nil
                            currentBandwidth = nil
                            currentVariantUrl = nil
                            currentWidth = 0
                            currentHeight = 0
                        }
                    }
                }
                
                DispatchQueue.main.async {
                    completion(variants)
                }
            } catch {
                print("Error parsing HLS playlist: \(error)")
                DispatchQueue.main.async {
                    completion([])
                }
            }
        }
    }
    
    private static func resolveUrl(baseUrl: URL, relativeUrl: String) -> String? {
        if relativeUrl.hasPrefix("http://") || relativeUrl.hasPrefix("https://") {
            return relativeUrl
        }
        
        return URL(string: relativeUrl, relativeTo: baseUrl)?.absoluteString
    }
    
    static func isHlsUrl(_ url: String) -> Bool {
        return url.hasSuffix(".m3u8") || url.contains(".m3u8")
    }
}

